### ISSUE TITLE
Streamlit support

### ISSUE THREAD

**Issue by ghilesmeddour** [predict-idlab/plotly-resampler#77]
Hi,

I tried to use the lib with Streamlit. The figure is displayed correctly but the resmapling does not seem to work. Is there a trick to make it work with Streamlit or is this not supported yet? If not, is there any way to support it?

Thanks,

Ghiles

**Comment by jvdd** [comment]
Hi @ghilesmeddour,

Thanks for submitting this issue! Can you provide some code as minimal example of your problem?

I've never used streamlit, but I'll gladly look into it! 😊

Cheers, Jeroen

**Comment by ghilesmeddour** [comment]
Hi @jvdd,

Thanks for your answer. Here is a simple example:

```python
# resampler_demo.py

import plotly.graph_objects as go; import numpy as np
from plotly_resampler import FigureWidgetResampler
import streamlit as st

x = np.arange(1_000_000)
noisy_sin = (3 + np.sin(x / 200) + np.random.randn(len(x)) / 10) * x / 1_000

fig = FigureWidgetResampler(go.Figure())
fig.add_trace(go.Scattergl(name='noisy sine', showlegend=True), hf_x=x, hf_y=noisy_sin)

st.plotly_chart(fig)
```

You can run it using:

```
$ streamlit run resampler_demo.py
```

Dependencies:
```
python=3.10.4
streamlit=1.10.0
plotly-resampler=0.7.1
```

**Comment by jvdd** [comment]
Hi @ghilesmeddour,

Thanks for providing this example, this got me up and running quickly! After some trial & error, I boxed (together with @jonasvdd) this minimal working example together

The trick to use streamlit with plotly-resampler is;
1) use `FigureResampler`
2) execute `.show_dash` in a (sub)process on a certain *port*
    (-> this step ensures that code after the `.show_dash` can be executed)
3) add the visualization url (referencing the *port*) as iframe component to streamlit

```py
import plotly.graph_objects as go; import numpy as np
from plotly_resampler import FigureWidgetResampler, FigureResampler

x = np.arange(1_000_000)
noisy_sin = (3 + np.sin(x / 200) + np.random.randn(len(x)) / 10) * x / 1_000

### 1. Use FigureResampler
fig = FigureResampler(default_n_shown_samples=2_000)
fig.add_trace(go.Scattergl(name='noisy sine', showlegend=True), hf_x=x, hf_y=noisy_sin)
fig.update_layout(height=700)

### 2. Run the visualization (which is a dash app) in a (sub)process on a certain port
from multiprocessing import Process
port = 9022
proc = Process(
    target=fig.show_dash, kwargs=dict(mode="external", port=port)
)
proc.start()

# Deleting the lines below this and running this file will result in a classic running dash app
# Note: for just a dash app it is not even necessary to execute .show_dash in a (sub)process

### 3. Add as iframe component to 
import streamlit.components.v1 as components
components.iframe(f'http://localhost:{port}', height=700)
```
---

I'm interested in hearing your input on this implementation! Is there a better / cleaner way to integrate a dash app in streamlit? 

I'm also wondering why you are using plotly with streamlit (where I know very little of) instead of dash? 
*(I believe both are similar in functionaliity, but streamlit seems more convenient to use for quick prototyping) 
However, if there are clear advantages of the one over the other, I might try to build an integration example how streamlit can use those advantages and add it to our [examples](https://github.com/predict-idlab/plotly-resampler/tree/main/examples) folder!* :smile:

**Comment by ghilesmeddour** [comment]
Thank you very much @jvdd,

Thanks for the code, I just tested it, it works fine 😊 It seems to me a reasonable way to integrate a Dash app in Streamlit. I think it can help others and definitely deserves to be in the examples folder 🤓

I wonder though if it would be possible to do without Dash altogether or if the implementation of this lib was very tied to Dash? Is this lib on Jupyter using Dash underhood?

For Streamlit, I started using it recently and I must say I really like it. It allows you to prototype and build data apps much faster than Dash, the result is cleaner with less effort, and it's generally easier to debug and maintain.

**Comment by jvdd** [comment]
Glad to hear this implementation works well for you @ghilesmeddour! I'll include this example in our examples folder :smile: 

To answer your question; this lib is basically an add-on to plotly that adds data aggregation fucntionality to plotly figures by connecting the figure (which is the front-end where the aggregated data is shown) with a back-end (where the data is stored) via callbacks (that perform the data aggregation).
This back-end can be either;
* a dash app (when using the `FigureResampler`)
* an ipykernel (when using `FigureWidgetResampler`)

As streamlit does not support FigureWidgets (https://discuss.streamlit.io/t/could-we-have-plotly-figurewidget-support/4271), I believe the only way to integrate (at the moment) is by incorporating a dash application in your streamlit app.

**Comment by ghilesmeddour** [comment]
This is very clear. Thank you @jvdd for your help.

**Comment by WilliamJudge94** [comment]
Hello, I am having an issue with this implementation. I followed the instructions, but I am only left with a plotly figure with no color. I would like to plot multiple lines in a single plot and be able to easily distinguish between them.  Is there a way to get the full version of the resampled plotly plot within streamlit?

**Comment by jvdd** [comment]
Hey @WilliamJudge94,

I was able to reproduce the issue you described.
After a quick search, I stumbled upon these two related issues;
- https://discuss.streamlit.io/t/streamlit-overrides-colours-of-plotly-chart/34943
- https://discuss.streamlit.io/t/black-and-white-colors-are-only-seen-when-plotly-is-downloaded-as-html-file-or-a-png-or-jpeg-file/37851

TLDR; streamlit changed some stuff w.r.t. plotly colors being loaded / handled in the front-end to better serve a dark-theme => this is thus more a plotly x streamlit issue than a plotly-resampler issue :upside_down_face: 

To overcome this issue, you should now add the following code to your script :arrow_down:
```py
import plotly.io as pio
pio.templates.default = "plotly"
```

I'll update the streamlit example with this.

Hope this helps!

Cheers, Jeroen
